File: Assets/Scripts/BoardManager.cs
using System.Collections.Generic;
using UnityEngine;
using Random = UnityEngine.Random;

public class BoardManager : MonoBehaviour
{

    public class Count
    {
        public int minimum;
        public int maximum;

        public Count(int min, int max)
        {
            minimum = min;
            maximum = max;
        }
    }

    public int columns = 8;
    public int rows = 8;
    public Count wallCount = new Count(5, 9);
    public Count foodCount = new Count(1, 5);
    public GameObject exit;
    public GameObject[] floorTiles;
    public GameObject[] wallTiles;
    public GameObject[] foodTiles;
    public GameObject[] enemyTiles;
    public GameObject[] outerWallTiles;

    private Transform boardHolder;
    private List<Vector3> gridPositions = new List<Vector3>();


    private void InitializeList()
    {
        gridPositions.Clear();

        for(var x = 1; x < columns - 1; x ++)
            for(var y = 1; y < rows - 1; y++)
            {
                gridPositions.Add(new Vector3(x, y, 0f));
            }
    }

    private void BoardSetup()
    {
        boardHolder = new GameObject("Board").transform;

        for (var x = -1; x < columns + 1; x++)
            for (var y = -1; y < rows + 1; y++)
            {
                var toInstantiate = floorTiles[Random.Range(0, floorTiles.Length)];

                if (x == -1 || x == columns || y == -1 || y == rows)
                    toInstantiate = outerWallTiles[Random.Range(0, outerWallTiles.Length)];

                var instance = Instantiate(toInstantiate, new Vector3(x, y, 0f), Quaternion.identity);
                instance.transform.SetParent(boardHolder);
            }
    }

    private Vector3 RandomPosition()
    {
        var randomIndex = Random.Range(0, gridPositions.Count);
        var randomPostion = gridPositions[randomIndex];
        gridPositions.Remove(randomPostion);

        return randomPostion;
    }

    private void LayoutObjectAtRandom(GameObject[] tileArray, int minimum, int maximum)
    {
        var objectCount = Random.Range(minimum, maximum + 1);
        for(var i = 0; i < objectCount; i++)
        {
            var randomPostion = RandomPosition();
            var tileChoice = tileArray[Random.Range(0, tileArray.Length)];
            Instantiate(tileChoice, randomPostion, Quaternion.identity);
        }
    }

    public void SetupScene(int level)
    {
        BoardSetup();
        InitializeList();
        LayoutObjectAtRandom(wallTiles, wallCount.minimum, wallCount.minimum);
        LayoutObjectAtRandom(foodTiles, foodCount.minimum, foodCount.minimum);

        var enemyCount = (int)Mathf.Log(level, 2f);
        LayoutObjectAtRandom(enemyTiles, enemyCount, enemyCount);
        Instantiate(exit, new Vector3(columns - 1, rows - 1, 0f), Quaternion.identity);
    }
}

File: Assets/Scripts/BoardManager.fs
namespace Unity_2DRoguelike

open System
open System.Collections.Generic
open UnityEngine

[<Serializable>]
type Count =

    val mutable public _min : int
    val mutable public _max : int

    new (min : int, max : int) =
        {
            _min = min
            _max = max
        }

type BoardManager =
    inherit MonoBehaviour 

    val mutable public _columns : int
    val mutable public _rows : int
    val mutable public _wallCount : Count
    val mutable public _foodCount : Count
    
    [<DefaultValue>] val mutable public _exit : GameObject
    [<DefaultValue>] val mutable public _floorTiles : GameObject[]
    [<DefaultValue>] val mutable public _wallTiles : GameObject[]
    [<DefaultValue>] val mutable public _foodTiles : GameObject[]
    [<DefaultValue>] val mutable public _enemyTiles : GameObject[]
    [<DefaultValue>] val mutable public _outerWallTiles : GameObject[]

    [<DefaultValue>] val mutable private _boardHolder : Transform
    val mutable private _gridPositions : List<Vector3>

    new () =
        {
            _columns = 8
            _rows = 8
            _wallCount = new Count(5, 9 )
            _foodCount = new Count(1, 5)

            _gridPositions = new List<Vector3>()
        }

    member private this.InitializeList() =
        this._gridPositions.Clear()

        for x in 1 .. this._columns - 2 do 
            for y in 1 .. this._rows - 2 do
                this._gridPositions.Add(new Vector3((float32 x), (float32 y), 0.0f))

    member private this.BoardSetup() =
        let board = new GameObject("Board")
        this._boardHolder <- board.transform

        for x in -1 .. this._columns do 
            for y in -1 .. this._rows do
                let mutable toInstantiate = this._floorTiles.[UnityEngine.Random.Range(0, this._floorTiles.Length)]

                if x = -1 || x = this._columns || y = -1 || y = this._rows
                    then toInstantiate <- this._outerWallTiles.[UnityEngine.Random.Range(0, this._outerWallTiles.Length)];

                let instance = Object.Instantiate<GameObject>(toInstantiate, new Vector3((float32 x),(float32 y), 0.0f), Quaternion.identity);
                instance.transform.SetParent(this._boardHolder)

    member private this.RandomPosition() =
        let randomIndex = UnityEngine.Random.Range(0, this._gridPositions.Count)
        let randomPostion = this._gridPositions.[randomIndex]
        this._gridPositions.Remove(randomPostion) |> ignore

        randomPostion

    member private this.LayoutObjectAtRandom(tileArray : GameObject[], minimum : int, maximum : int) =
        let objectCount = UnityEngine.Random.Range(minimum, maximum + 1)
        
        for i in 0 .. objectCount do
            let randomPostion = this.RandomPosition()
            let tileChoice = tileArray.[UnityEngine.Random.Range(0, tileArray.Length)]
            Object.Instantiate(tileChoice, randomPostion, Quaternion.identity) |> ignore

    member public this.SetupScene(level : int) = 
        this.BoardSetup()
        this.InitializeList()
        this.LayoutObjectAtRandom(this._wallTiles, this._wallCount._min, this._wallCount._max)
        this.LayoutObjectAtRandom(this._foodTiles, this._foodCount._min, this._foodCount._max)

        let enemyCount = Mathf.Log(float32 level, 2.0f)
        this.LayoutObjectAtRandom(this._enemyTiles, (int enemyCount), (int enemyCount))
        let position = new Vector3(this._columns - 1 |> float32, this._rows - 1 |> float32, 0.0f)
        Object.Instantiate<GameObject>(this._exit, position, Quaternion.identity) |> ignore